lm-simulator/lm_simulator/crud.py

```diff
@@ -1,129 +1,155 @@
-    a LicenseNotFound exception.
-    Given that the license exists in the database, the total number of licenses must be greater than or equal
-    the in_use value for the license plus the quantity needed for the new license_in_use. If it is not, then
-    raise a NotEnoughLicenses exception.
-    If all the conditions are satisfied, the license_in_use is created and returned.
+async def add_license_in_use(session: Session, license_in_use: LicenseInUseCreate) -> LicenseInUseRow:
     """
-    license = _get_license_available(session, license_in_use.license_name)
-    if not license:
-        raise LicenseNotFound()
-    license_row = LicenseRow.from_orm(license)
-    if license_row.total < license_in_use.quantity + license_row.in_use:
-        raise NotEnoughLicenses()
-    db_license_in_use = LicenseInUse(**license_in_use.dict())
-    session.add(db_license_in_use)
-    session.commit()
-    session.refresh(db_license_in_use)
-    return LicenseInUseRow.from_orm(db_license_in_use)
-
-
-def _get_licenses_in_database(
-    session: Session,
-    lead_host: str,
-    user_name: str,
-    quantity: int,
-    license_name: str,
-) -> List[LicenseInUse]:
-    stmt = (
-        select(LicenseInUse)
-        .join(License)
-        .where(
-            License.name == license_name,
-            LicenseInUse.lead_host == lead_host,
-            LicenseInUse.user_name == user_name,
-            LicenseInUse.quantity == quantity,
-        )
+    Add a new LicenseInUse to the database.
+    """
+    query = await session.execute(select(License).where(License.name == license_in_use.license_name))
+    db_license = query.scalar_one_or_none()
+    enforce_defined(
+        db_license,
+        "License not found",
+        raise_exc_class=HTTPException,
+        exc_builder=lambda exc_class, msg: exc_class(
+            status_code=status.HTTP_404_NOT_FOUND,
+            detail="License not found",
+        ),
+    )
+    license_row = LicenseRow.model_validate(db_license)
+
+    require_condition(
+        license_row.total >= license_in_use.quantity + license_row.in_use,
+        "Not enough licenses",
+        raise_exc_class=HTTPException,
+        exc_builder=lambda exc_class, msg: exc_class(
+            status_code=status.HTTP_400_BAD_REQUEST,
+            detail="Not enough licenses",
+        ),
     )
-    licenses = session.execute(stmt).scalars().all()
-    return licenses
 
+    db_license_in_use = LicenseInUse(**license_in_use.model_dump())
 
-def delete_license_in_use(
-    session: Session,
-    lead_host: str,
-    user_name: str,
-    quantity: int,
-    license_name: str,
-):
-    """Delete the license_in_use from the database.
+    with handle_errors(
+        "Can't create License In Use",
+        raise_exc_class=HTTPException,
+        exc_builder=lambda exc_class, msg: exc_class(
+            status_code=status.HTTP_400_BAD_REQUEST,
```

Comment:
Same comment as above. I don't think 400 is the right error code for this.

Reply:
Whe you can't create, it means that either the database operation went wrong or the license is already fully in use.
Is it possible to have a different treatment depending on the error returned by the database?

Reply:
Since we're checking if there's enough licenses before inserting, that makes sense that an exception here means that something went wrong in the database. I updated it to 500.